server/sonar-server/src/main/java/org/sonar/server/issue/ws/ChangelogAction.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.issue.ws;
+
+import com.google.common.io.Resources;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import org.sonar.api.server.ws.Request;
+import org.sonar.api.server.ws.Response;
+import org.sonar.api.server.ws.WebService;
+import org.sonar.core.issue.FieldDiffs;
+import org.sonar.core.util.stream.Collectors;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.issue.IssueDto;
+import org.sonar.db.user.UserDto;
+import org.sonar.server.issue.IssueFinder;
+import org.sonarqube.ws.Issues.ChangelogWsResponse;
+
+import static com.google.common.base.Strings.emptyToNull;
+import static org.sonar.api.utils.DateUtils.formatDateTime;
+import static org.sonar.core.util.Protobuf.setNullable;
+import static org.sonar.core.util.Uuids.UUID_EXAMPLE_01;
+import static org.sonar.server.issue.IssueFieldsSetter.TECHNICAL_DEBT;
+import static org.sonar.server.ws.WsUtils.writeProtobuf;
+import static org.sonarqube.ws.client.issue.IssuesWsParameters.PARAM_ISSUE;
+
+public class ChangelogAction implements IssuesWsAction {
+
+  private static final String ACTION_CHANGELOG = "changelog";
+  private static final String EFFORT_CHANGELOG_KEY = "effort";
+
+  private final DbClient dbClient;
+  private final IssueFinder issueFinder;
+
+  public ChangelogAction(DbClient dbClient, IssueFinder issueFinder) {
+    this.dbClient = dbClient;
+    this.issueFinder = issueFinder;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction(ACTION_CHANGELOG)
+      .setDescription("Display changelog of an issue")
+      .setSince("4.1")
+      .setHandler(this)
+      .setResponseExample(Resources.getResource(IssuesWs.class, "example-changelog.json"));
+    action.createParam(PARAM_ISSUE)
+      .setDescription("Key of the issue")
+      .setRequired(true)
+      .setExampleValue(UUID_EXAMPLE_01);
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
```

Comment:
I have the feeling you could go further with the functional style and improving the readability